### ISSUE TITLE
Host filter doesn't exactly match

### ISSUE THREAD

**Issue by cseeger-epages** [capistrano/capistrano#1766]
Greedy host filter
---

#### Steps to reproduce
eg. cap stage --hosts=system01 <some task>
using a stage with defined hosts:
system011
system012
system020

#### Expected behaviour

empty host list or ERROR Host does not exists or sth. like this

#### Actual behaviour

matching any host starting with the hosts parameter

in this case
system011 and system012

#### Fix

from: https://github.com/capistrano/capistrano/blob/master/lib/capistrano/configuration/host_filter.rb#L17-L27
```ruby
def regex_matcher(values)
        values.map! do |v|
          case v
          when Regexp then v
          else
            vs = v.to_s
            vs =~ /^[-A-Za-z0-9.]+$/ ? vs : Regexp.new(vs)
          end
        end
        Regexp.union values
      end
```

Line 23 needs to be changed from
```ruby
vs =~ /^[-A-Za-z0-9.]+$/ ? vs : Regexp.new(vs)
```
to
```ruby
vs =~ /^[-A-Za-z0-9.]+$/ ? %r{^#{vs}$} : Regexp.new(vs)
```

#### Addition

Wildcard matching should be used while directly using a regex in capistrano 

eg. cap stage --hosts='system01\d' <some task>

using only a string it should match exactly

**Comment by mattbrictson** [comment]
Yes, I believe this is a bug. The [documentation](http://capistranorb.com/documentation/advanced-features/host-filtering/) implies that the strings should be matched exactly, unless certain regular expression syntax is used.

@leehambley @will-in-wi @cshaffer Do you agree this is a bug?

---

Also, I don't think `%r{^#{vs}$}` is exactly correct, since a `.` in the filter string will be interpreted as a matching any single character. Probably what we want is `%r{^#{Regexp.quote(vs)}$}`?

**Comment by will-in-wi** [comment]
I agree that this appears to be a bug. Unless @cseeger-epages is interested in doing so, I can probably whip up a PR.

**Comment by mattbrictson** [comment]
@will-in-wi Thanks. Can you take the lead on seeing this issue is ultimately resolved? I'll assign it to you if that's OK.

**Comment by will-in-wi** [comment]
Sounds good to me. @cseeger-epages: Are you interested in making a pull request for this? You already found the offending line. The only additional thing I would suggest would be to make an rspec test that replicates this issue, and another one which checks for the case when what is passed contains a dot.

**Comment by cseeger-epages** [comment]
We tested the better solution from @mattbrictson
```ruby
[1] pry(main)> vs = "wl"
=> "wl"
[2] pry(main)> %r{^#{Regexp.quote(vs)}$}
=> /^wl$/
[3] pry(main)> vs = "wl.example.com"
=> "wl.example.com"
[4] pry(main)> %r{^#{Regexp.quote(vs)}$}
=> /^wl\.example\.com$/
```
i will make a pull request during the day for this. Thanks for your fast replys.

Additionally:
The same is code is used in the roles_filter:
https://github.com/capistrano/capistrano/blob/master/lib/capistrano/configuration/role_filter.rb#L17-L27
```ruby
def regex_matcher(values)
        values.map! do |v|
          case v
          when Regexp then v
          else
            vs = v.to_s
            vs =~ %r{^/(.+)/$} ? Regexp.new($1) : /^#{vs}$/
          end
        end
        Regexp.union values
      end
```

should this also be fixed to ```%r{^#{Regexp.quote(vs)}$}``` ? if someone wants to define some roles using a ```.```  maybe "somerole.somesubrole"

**Comment by will-in-wi** [comment]
@mattbrictson might weigh in here, but I'd say that the roles_filter fix makes sense too. If you want to make that PR, could you do it as an additional one?

Thanks!

**Comment by cseeger-epages** [comment]
Ok if @mattbricson gives his ok i will make an additional one.

**Comment by mattbrictson** [comment]
Yes, I'll accept a PR for fixing the roles filter as well. Thanks!

**Comment by cseeger-epages** [comment]
should i open a new issue for this or is it ok to to reference it here ?

**Comment by will-in-wi** [comment]
I'd just push the PR and reference it here. If @mattbrictson wants a new issue, then it can be created at that point.

**Comment by cseeger-epages** [comment]
Ok thx. If both PR's get merged can someone fix the changelog, theres no need for 2 lines for the same issue.

**Comment by mattbrictson** [comment]
This can be closed after #1771 is merged (once merge conflicts are fixed).

**Comment by mattbrictson** [comment]
Fixed via #1771
